pymc4/flow/evaluate.py

```diff
@@ -0,0 +1,171 @@
+                # for some reason outer scope (control flow) may silence an exception in `yield`
+                # try:
+                #     yield something_bad
+                # except:
+                #     pass
+                # ...
+                # in that case `gen.throw(error)` will go silently, but we can keep track of it and capture
+                # silent behaviour and create an error in an model evaluation scope, not in control flow
+                # this error handling should prevent undefined behaviour. However, if the control flow does not have
+                # error handling stuff, the Exception is injected directly to the place where it was occurred
+                # if we ever appear at this particular point we may by whatever deep in call stack of `evaluate_model`
+                # and therefore bypass this error up the stack
+                # -----
+                # this message with little modifications will appear in exception
+                control_flow.close()
+                raise
+            except EarlyReturn as e:
+                # for some reason we may raise it within model evaluation,
+                # e.g. in self.proceed_distribution
+                return e.args
+            except StopIteration as stop_iteration:
+                return_value, state = self.finalize_control_flow(stop_iteration, model_info, state)
+                break
+        return return_value, state
+
+
+SamplingState = collections.namedtuple("EvaluationState", "values,distributions")
+
+
+class SamplingExecutor(Executor):
+    def __init__(self, **do):
+        self.default_do = do
+
+    def new_state(self, **do):
+        state_do = self.default_do.copy()
+        state_do.update(do)
+        return SamplingState(values=state_do, distributions=dict())
+
+    def modify_distribution(self, dist, model_info, state):
+        return dist
+
+    def proceed_distribution(self, dist, model_info, state):
+        scoped_name = pymc4.scopes.Scope.variable_name(dist.name)
+        if scoped_name in state.distributions:
+            raise EvaluationError(
+                "Attempting to create duplicate variable '{}', "
+                "this may happen if you forget to use `pm.name_scope()` when calling same "
+                "model/function twice without providing explicit names. If you see this "
+                "error message and the function being called is not wrapped with "
+                "`pm.model`, you should better wrap it to provide explicit name for this model"
+                .format(scoped_name))
+        if scoped_name in state.values:
+            return_value = state.values[scoped_name]
+        else:
+            return_value = state.values[scoped_name] = dist.sample()
+        state.distributions[scoped_name] = dist
+        return return_value, state
+
+    def prepare_model_control_flow(self, model, model_info, state):
+        control_flow: types.GeneratorType = model.control_flow()
```

Comment:
This is pretty magically to me... what does this line do?

Reply:
model you see in arguments keeps a no-arg function that creates a generator that yields distributions. I name that generator `control_flow`. For some reason, my type checker was silly to detect generator type here, I helped it

Reply:
Oh so the `: types.GeneratorType` part in `control_flow: types.GeneratorType = model.control_flow()` is type hint right?

Reply:
Right

Reply:
Minor grammar correction:

> "this may happen if you forget to use `pm.name_scope()` when calling same "

should be 

> "this may happen if you forget to use `pm.name_scope()` when calling the same "